File: SwapNodesInPairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        res = ListNode(0)
        res.next = head
        if res.next is None or res.next.next is None:
            return res.next
        p = res
        while p.next and p.next.next:
            tmp = p.next.next
            p.next.next = tmp.next
            tmp.next = p.next
            p.next = tmp
            p = p.next.next
                
        return res.next

if __name__ == '__main__':
    h = ListNode(1)
    h.next = ListNode(2)
    so = Solution()
    res =  so.swapPairs(h)
    while res:
        print res.val
        res = res.next
    

File: CPP/Second/swap_nodes_in_pairs.cc
// https://oj.leetcode.com/problems/swap-nodes-in-pairs/

// Given a linked list, swap every two adjacent nodes and return its head.
// For example,
// Given 1->2->3->4, you should return the list as 2->1->4->3.
// Your algorithm should use only constant space. You may not modify the values in

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        ListNode **p = &head;
        while ((*p) && (*p)->next) {
            ListNode *tmp = (*p);
            (*p) = (*p)->next;
            tmp->next = (*p)->next;
            (*p)->next = tmp;

            p = &(*p)->next->next;
        }
        return head;
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
